javatests/google/registry/rde/XjcToDomainResourceConverterTest.java

```diff
@@ -0,0 +1,484 @@
+        try (InputStream ins = new ByteArrayInputStream(entry.getXmlBytes())) {
+          XjcRdeDomain unmarshalledXml =
+              ((XjcRdeDomainElement) unmarshaller.unmarshal(ins)).getValue();
+          assertThat(unmarshalledXml.getName()).isEqualTo(xjcDomain.getName());
+          assertThat(unmarshalledXml.getRoid()).isEqualTo(xjcDomain.getRoid());
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      }
+    });
+  }
+
+  @Test
+  public void testConvertDomainResourceAutoRenewBillingEvent() throws Exception {
+    persistActiveContact("jd1234");
+    persistActiveContact("sh8013");
+    final XjcRdeDomain xjcDomain = loadDomainFromRdeXml("domain_fragment.xml");
+    // First import in a transaction,
+    // then verify in another transaction.
+    // Ancestor queries don't work within
+    // the same transaction.
+    ofy().transact(new VoidWork() {
+      public void vrun() {
+        DomainResource domain = XjcToDomainResourceConverter.convertDomain(xjcDomain);
+        ofy().save().entity(domain);
+      }
+    });
+    ofy().transact(new VoidWork() {
+      public void vrun() {
+        List<HistoryEntry> historyEntries = ofy().load().type(HistoryEntry.class)
+            .ancestor(Key.create(DomainResource.class, xjcDomain.getRoid())).list();
+        assertThat(historyEntries.size()).isEqualTo(1);
+        HistoryEntry entry = historyEntries.get(0);
+        List<BillingEvent.Recurring> billingEvents = ofy().load().type(BillingEvent.Recurring.class)
+            .ancestor(entry).list();
+        assertThat(billingEvents.size()).isEqualTo(1);
+        BillingEvent.Recurring autoRenewEvent = billingEvents.get(0);
+        assertThat(autoRenewEvent.getReason()).isEqualTo(Reason.RENEW);
+        assertThat(autoRenewEvent.getFlags()).isEqualTo(ImmutableSet.of(Flag.AUTO_RENEW));
+        assertThat(autoRenewEvent.getTargetId()).isEqualTo(xjcDomain.getRoid());
+        assertThat(autoRenewEvent.getClientId()).isEqualTo(xjcDomain.getClID());
+        assertThat(autoRenewEvent.getEventTime()).isEqualTo(xjcDomain.getExDate());
+        assertThat(autoRenewEvent.getRecurrenceEndTime()).isEqualTo(END_OF_TIME);
+      }
+    });
+  }
+
+  @Test
+  public void testConvertDomainResourceAutoRenewPollMessage() throws Exception {
+    persistActiveContact("jd1234");
+    persistActiveContact("sh8013");
+    final XjcRdeDomain xjcDomain = loadDomainFromRdeXml("domain_fragment.xml");
+    // First import in a transaction,
+    // then verify in another transaction.
+    // Ancestor queries don't work within
+    // the same transaction.
+    ofy().transact(new VoidWork() {
+      public void vrun() {
+        DomainResource domain = XjcToDomainResourceConverter.convertDomain(xjcDomain);
+        ofy().save().entity(domain);
+      }
+    });
+    ofy().transact(new VoidWork() {
+      public void vrun() {
+        List<HistoryEntry> historyEntries = ofy().load().type(HistoryEntry.class)
```

Comment:
There's a lot of useful helper methods in DatastoreHelper that you can reduce a lot of this boilerplate.  I recommend looking through that class just so you're aware of what's already implemented, and then use the ones that are relevant.

For example, this could more easily be handled by using `DatastoreHelper.getHistoryEntries()`